File: python/examples/10_custom_policy.py
"""
Parmana Python SDK

Example 10

Custom Policy

Demonstrates selecting an explicit policy version
for execution.

The BusinessTransaction specifies the PolicyReference.

Parmana Runtime loads exactly that policy.

It never discovers or selects policies automatically.

Authority
        ↓
Authorization
        ↓
Intent
        ↓
Business Transaction
        ↓
PolicyReference
        ↓
Policy Engine
        ↓
Decision
        ↓
Execution
        ↓
Verification
        ↓
Receipt
"""

from datetime import datetime
from uuid import uuid4

from parmana import ParmanaClient

from parmana.models.authority import Authority
from parmana.models.authorization import Authorization
from parmana.models.business_transaction import BusinessTransaction
from parmana.models.decision import (
    Decision,
    DecisionOutcome,
)
from parmana.models.execution import (
    Execution,
    ExecutionMode,
    ExecutionStatus,
)
from parmana.models.execution_trust_record import (
    ExecutionTrustRecord,
)
from parmana.models.intent import Intent
from parmana.models.policy_reference import PolicyReference
from parmana.models.receipt import Receipt
from parmana.models.verification import (
    Verification,
    VerificationStatus,
)


def main() -> None:

    #
    # Authority
    #

    authority = Authority(
        authority_id=str(uuid4()),
        authority_type="ENTERPRISE",
        authority_name="Acme Manufacturing",
        created_at=datetime.utcnow(),
        public_key="enterprise-public-key",
        signature_algorithm="Ed25519",
    )

    #
    # Authorization
    #

    authorization = Authorization(
        authorization_id=str(uuid4()),
        authority_id=authority.authority_id,
        subject="robot-arm-07",
        action="START_PRODUCTION",
        resource="assembly-line-3",
        issued_at=datetime.utcnow(),
    )

    #
    # Intent
    #

    intent = Intent(
        intent_id=str(uuid4()),
        operation="START_PRODUCTION",
        target="assembly-line-3",
        parameters={
            "batch": "BATCH-2026-001",
            "quantity": 500,
        },
    )

    #
    # Explicit Policy Reference
    #

    policy = PolicyReference(
        policy_name="manufacturing-production-policy",
        policy_version="2.3.1",
    )

    #
    # Business Transaction
    #

    transaction = BusinessTransaction(
        business_transaction_id=str(uuid4()),
        authority=authority,
        authorization=authorization,
        intent=intent,
        policy_reference=policy,
        created_at=datetime.utcnow(),
    )

    #
    # Runtime evaluates EXACTLY this policy.
    #

    decision = Decision(
        decision_id=str(uuid4()),
        intent_id=intent.intent_id,
        policy=policy,
        signals={
            "machine_ready": True,
            "operator_present": True,
            "quality_checks_passed": True,
            "emergency_stop": False,
        },
        outcome=DecisionOutcome.APPROVED,
        reason="Policy evaluation approved.",
        evaluated_at=datetime.utcnow(),
    )

    execution = Execution(
        execution_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        decision=decision,
        status=ExecutionStatus.COMPLETED,
        mode=ExecutionMode.SYNC,
        started_at=datetime.utcnow(),
        completed_at=datetime.utcnow(),
        evidence={
            "policy_loaded": policy.policy_name,
            "policy_version": policy.policy_version,
            "production_started": True,
        },
    )

    verification = Verification(
        verification_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        status=VerificationStatus.VERIFIED,
        message="Execution verified.",
        verified_at=datetime.utcnow(),
        trust_record_hash="production-trust-record",
    )

    receipt = Receipt(
        receipt_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        execution_id=execution.execution_id,
        trust_record_hash="production-trust-record",
        receipt_hash="receipt-hash",
        signature="enterprise-signature",
        algorithm="Ed25519",
        issued_at=datetime.utcnow(),
    )

    record = ExecutionTrustRecord(
        trust_record_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        transaction=transaction,
        overrides=(),
        executions=(execution,),
        verifications=(verification,),
        receipts=(receipt,),
        trust_record_hash="production-trust-record",
        created_at=datetime.utcnow(),
        updated_at=datetime.utcnow(),
    )

    client = ParmanaClient()

    verification = client.verify(record)

    print()
    print("==========================================")
    print("Parmana Custom Policy")
    print("==========================================")

    print(f"Authority        : {authority.authority_name}")
    print(f"Operation        : {intent.operation}")

    print()

    print("Policy")

    print("--------------------------------")

    print(f"Name             : {policy.policy_name}")
    print(f"Version          : {policy.policy_version}")

    print()

    print("Runtime")

    print("--------------------------------")

    print("Policy Selection : Explicit")
    print("Policy Discovery : Disabled")
    print("Decision         : APPROVED")

    print()

    print("Execution")

    print("--------------------------------")

    print(f"Status           : {execution.status.value}")
    print(f"Mode             : {execution.mode.value}")

    print()

    print("Verification")

    print("--------------------------------")

    print(f"Status           : {verification.status.value}")

    print()

    print("Canonical Guarantee")

    print("--------------------------------")

    print(
        "The BusinessTransaction specifies the "
        "PolicyReference."
    )

    print(
        "The Runtime loads exactly that policy."
    )

    print(
        "The Runtime never discovers or chooses "
        "another policy."
    )

    print()

    print(
        "Execution completed with a deterministic "
        "policy selection."
    )

    print("==========================================")


if __name__ == "__main__":
    main()

File: typescript/examples/10_custom_policy.ts
/**
 * Parmana TypeScript SDK
 *
 * Example 10
 *
 * Custom Policy
 *
 * Demonstrates explicit PolicyReference selection.
 *
 * The BusinessTransaction specifies exactly which
 * policy version must be evaluated.
 *
 * The Runtime never discovers or chooses policies.
 */

import {
  Authority,
  Authorization,
  BusinessTransaction,
  DecisionOutcome,
  Execution,
  ExecutionMode,
  ExecutionStatus,
  ExecutionTrustRecord,
  Intent,
  ParmanaClient,
  PolicyReference,
} from "../src/index.js";

async function main(): Promise<void> {
  const client = new ParmanaClient({
    endpoint: "http://localhost:8080",
  });

  const authority: Authority = {
    authorityId: "authority-001",
    authorityName: "Acme Manufacturing",
    createdAt: new Date(),
  };

  const authorization: Authorization = {
    authorizationId: "authorization-001",
    authorityId: authority.authorityId,
    subject: "manufacturing-controller",
    permissions: [
      "START_PRODUCTION",
    ],
    issuedAt: new Date(),
    expiresAt: new Date(
      Date.now() + 365 * 24 * 60 * 60 * 1000,
    ),
  };

  const intent: Intent = {
    intentId: "intent-001",
    authorizationId: authorization.authorizationId,
    operation: "START_PRODUCTION",
    target: "BATCH-2026-001",
    createdAt: new Date(),
  };

  /**
   * Explicit policy selection.
   *
   * This policy becomes part of the immutable
   * execution trust chain.
   */
  const policy: PolicyReference = {
    policyName: "manufacturing-production-policy",
    policyVersion: "2.3.1",
  };

  const transaction: BusinessTransaction = {
    businessTransactionId: "production-001",
    authority,
    authorization,
    intent,
    policy,
    createdAt: new Date(),
  };

  const execution: Execution = {
    executionId: "execution-001",

    businessTransactionId:
      transaction.businessTransactionId,

    decision: {
      decisionId: "decision-001",

      intentId: intent.intentId,

      policy,

      signals: {
        machineReady: true,
        operatorCertified: true,
        safetyInspectionPassed: true,
        inventoryAvailable: true,
      },

      outcome: DecisionOutcome.APPROVED,

      reason:
        "Production policy v2.3.1 approved execution.",

      evaluatedAt: new Date(),
    },

    status: ExecutionStatus.COMPLETED,

    mode: ExecutionMode.SYNC,

    startedAt: new Date(),

    completedAt: new Date(),

    evidence: {
      productionLine: "LINE-3",
      batch: "BATCH-2026-001",
      productionStarted: true,
    },
  };

  const trustRecord: ExecutionTrustRecord = {
    trustRecordId: "trust-record-001",

    businessTransactionId:
      transaction.businessTransactionId,

    transaction,

    overrides: [],

    executions: [
      execution,
    ],

    verifications: [],

    receipts: [],

    trustRecordHash:
      "91c84e64b42b4fa89b2df6f7a37fd0d5",

    createdAt: new Date(),

    updatedAt: new Date(),
  };

  console.log();
  console.log("======================================");
  console.log(" Parmana TypeScript SDK");
  console.log(" Example 10 - Custom Policy");
  console.log("======================================");
  console.log();

  console.log("Business Transaction");
  console.log("------------------------------");
  console.log(
    "Transaction :",
    transaction.businessTransactionId,
  );
  console.log(
    "Operation   :",
    intent.operation,
  );
  console.log(
    "Target      :",
    intent.target,
  );

  console.log();

  console.log("Explicit Policy");
  console.log("------------------------------");
  console.log(
    "Policy Name :",
    policy.policyName,
  );
  console.log(
    "Version     :",
    policy.policyVersion,
  );

  console.log();

  console.log("Decision");
  console.log("------------------------------");
  console.log(
    "Outcome      :",
    execution.decision.outcome,
  );
  console.log(
    "Reason       :",
    execution.decision.reason,
  );

  console.log();

  console.log("Runtime Signals");
  console.log("------------------------------");
  console.log(
    execution.decision.signals,
  );

  console.log();

  console.log("Execution");
  console.log("------------------------------");
  console.log(
    "Status       :",
    execution.status,
  );
  console.log(
    "Mode         :",
    execution.mode,
  );
  console.log(
    "Evidence     :",
    execution.evidence,
  );

  console.log();

  console.log("Execution Trust Record");
  console.log("------------------------------");
  console.log(
    "Trust Record :",
    trustRecord.trustRecordId,
  );
  console.log(
    "Hash         :",
    trustRecord.trustRecordHash,
  );

  console.log();

  console.log(
    "Submitting transaction to Parmana Runtime...",
  );

  try {
    const receipt = await client.execute(
      transaction,
    );

    console.log();

    console.log("Receipt");
    console.log("------------------------------");
    console.log(
      "Receipt ID   :",
      receipt.receiptId,
    );
    console.log(
      "Algorithm    :",
      receipt.algorithm,
    );
    console.log(
      "Issued At    :",
      receipt.issuedAt,
    );
  } catch (error) {
    console.log();
    console.log(
      "Runtime not available (expected during SDK development).",
    );

    if (error instanceof Error) {
      console.log(error.message);
    }
  }

  console.log();

  console.log("Execution Trust Chain");
  console.log("------------------------------");
  console.log("✓ Authority");
  console.log("✓ Authorization");
  console.log("✓ Intent");
  console.log("✓ Explicit PolicyReference");
  console.log("✓ Policy Evaluation");
  console.log("✓ Decision");
  console.log("✓ Execution");
  console.log("✓ Execution Evidence");
  console.log("✓ Execution Trust Record");

  console.log();

  console.log("Architectural Principle");
  console.log("------------------------------");
  console.log(
    "BusinessTransaction specifies the PolicyReference.",
  );
  console.log(
    "The Runtime loads exactly one policy version.",
  );
  console.log(
    "The Runtime never discovers or selects policies.",
  );
  console.log(
    "Deterministic replay always evaluates",
  );
  console.log(
    "the same policy version.",
  );

  console.log();

  console.log("Example completed.");
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});